utils.py

```diff
@@ -0,0 +1,188 @@
+            # no one anywhere got this award. No sense wasting resources on it.
+            try:
+                logging.info(f"No data in {awards.loc[awards.id == idx, 'name'].values[0]} for {row.paxminer_schema}")
+            except IndexError:
+                logging.error(f"{row.paxminer_schema} doesn't have achievement {idx} in their awards_list table.")
+            continue
+        new_data = _check_for_new_results(row, year, idx, df, awarded)
+        if new_data.empty:
+            # there is data but nothing new since the last run. Carry on.
+            try:
+                logging.info(
+                    f"{row.paxminer_schema} has data but nothing new for {awards.loc[awards.id == idx, 'name'].values[0]}."
+                )
+            except IndexError:
+                logging.error(f"{row.paxminer_schema} has new data but doesn't track achievement_id {idx}.")
+            continue
+
+        # we got this far so there are achievements to award.
+        if idx not in awards.id.tolist():
+            logging.error(f"{row.paxminer_schema} doesn't track achievement_id {idx}.")
+            continue
+
+        # Loop over each record in `new_data`, assiging as appropriate
+        for record in new_data.itertuples(index=False):
+            _d[record.pax_id].update({idx: 1})
+            new_award_name = awards.query("id == @idx")["name"].item()
+            new_award_verb = awards.query("id == @idx")["verb"].item()
+            total_achievements = _d[record.pax_id].total()
+            total_idx_achievements = _d[record.pax_id][idx]
+            ending = ordinal_suffix(total_idx_achievements)
+
+            sMessage = [
+                f"Congrats to our man <@{record.pax_id}>! ",
+                f"He just unlocked the achievement *{new_award_name}* for {new_award_verb}. ",
+                f"This is achievement #{total_achievements} for <@{record.pax_id}> and the {total_idx_achievements}{ending} ",
+                "time this year he's earned this award. Keep up the good work!",
+            ]
+            sMessage = "".join(sMessage)
+            logging.info(sMessage)
+            try:
+                response = client.chat_postMessage(channel=row.achievement_channel, text=sMessage, link_names=True)
+                logging.info(f"Successfully sent slack message for {record.pax_id} and achievement {idx}")
+            except SlackApiError as e:
+                if e.response.status_code == 429:
+                    delay = int(e.response.headers['Retry-After'])
+                    logging.info(f"Pausing Slack notifications for {delay} seconds.")
+                    time.sleep(delay)
+                    response = client.chat_postMessage(channel=row.achievement_channel, text=sMessage, link_names=True)
+                    logging.info(f"Successfully sent slack message for {record.pax_id} and achievement {idx}")
+            client.reactions_add(channel=row.achievement_channel, name="fire", timestamp=response["ts"])
```

Comment:
There can be other reasons the posting can fail, in which case this reaction add will bomb your run

Reply:
Sounds like Slack API lacks something to be desired. I'll wrap the reaction in a "finally" for now and look at the API for other status codes to check for. Added some logic that in the event error code 429 is _not_ the error code, capture it, log it and continue without posting the reaction. Hopefully this will stop the issue while letting me be proactive about fixing other issues as they pop up.